### PR TITLE
adding more info to the API reference page

### PR DESCRIPTION
**Pull request by d45**
This PR is to add more information about the TSC library API methods and classes.  You can preview the updated docs here:  https://d45.github.io/server-client-python/docs/api-ref 

The PR also includes an update to the navigation pane (docs_menu.html) for the new sections.  Looks like there is a minor merge conflict with that file (spaces or something). 

 
NOTE:  Merging into master so that the update goes live when the PR is approved.  
I could also merge into the Dev branch, if you prefer.  That would require another move to get the files to master. Squash and merge if it seems OK after review.   Thanks.

### CODE REVIEW

**Comment by d45** [comment]
OK. I spoke too soon. The api-ref.md file has some conflicts. The new version should replace the old.

File: docs/docs/api-ref.md [hunk 1]
```diff
@@ -5,257 +5,3008 @@ layout: docs
 
 <div class="alert alert-info">
     <b>Important:</b> More coming soon! This section is under active construction and might not reflect all the available functionality of the TSC library.
-    Until this reference is completed, we have noted the source files in the TSC library where you can get more information for individual endpoints.
-</div>
+    <!--- Until this reference is completed, we have noted the source files in the TSC library where you can get more information for individual endpoints. --->
```
Comment (RussTheAerialist) [inline docs/docs/api-ref.md]: Just delete the comment.  We have version control if we need to revive this line.

File: docs/docs/api-ref.md [hunk 2]
```diff
@@ -5,257 +5,3008 @@ layout: docs
...
+
+The TSC API reference is organized by resource. The TSC library is modeled after the REST API. The methods, for example, `workbooks.get()`, correspond to the endpoints for resources, such as [workbooks](#workbooks), [users](#users), [views](#views), and [data sources](#data-sources). The model classes (for example, the [WorkbookItem class](#workbookitem-class) have attributes that represent the fields (`name`, `id`, `owner_id`) that are in the REST API request and response packages, or payloads. 
+
+|:---  |  
+| **Note:**  Some methods and features provided in the REST API might not be currently available in the TSC library.  In addition, the same limitations apply to the TSC library that apply to the REST API with respect to resources on Tableau Server and Tableau Online. For more information, see the [Tableau Server REST API Reference](http://onlinehelp.tableau.com/current/api/rest_api/en-us/help.htm#REST/rest_api_ref.htm#API_Reference%3FTocPath%3DAPI%2520Reference%7C_____0){:target="_blank"}.|
```
Comment (RussTheAerialist) [inline docs/docs/api-ref.md]: I would add "and vice-versa" because we have functionality in TSC that isn't released to the public yet (though they are in betas)

File: docs/docs/api-ref.md [hunk 3]
```diff
@@ -5,257 +5,3008 @@ layout: docs
...
+
+ 
+
+* TOC  
+{:toc }
+<!-- {:toc ::options toc_levels="1,2"} -->
```
Comment (RussTheAerialist) [inline docs/docs/api-ref.md]: delete comments

File: docs/docs/api-ref.md [hunk 4]
```diff
@@ -5,257 +5,3008 @@ layout: docs
...
+<br>  
 
-Signs you in to Tableau Server.
+### TableauAuth class
 
 ```py
-Auth.sign_in(authentication_object)
+TableauAuth(username, password, site_id='', user_id_to_impersonate=None)
 ```
+The `TableauAuth` class contains the attributes for the authentication resources. The `TableauAuth` class defines the information you can set in a  request or query from Tableau Server. The class members correspond to the attributes of a server request or response payload. To use this class, create a new instance, supplying user name, password, and site information if necessary, and pass the request object to the [Auth.sign_in](#auth.sign-in) method.
```
Comment (RussTheAerialist) [inline docs/docs/api-ref.md]: Probably include a note about other authentication schemes may be added in the future (e.g. oauth)

File: docs/docs/api-ref.md [hunk 5]
```diff
@@ -5,257 +5,3008 @@ layout: docs
...
 
-### Sign Out
+**Attributes**  
 
-Signs you out of Tableau Server.
+Name | Description   
+:--- | :---   
+`username` | The name of the user whose credentials will be used to sign in.   
+`password` | The password of the user.     
+`site_id` | This corresponds to the `contentUrl` attribute in the Tableau REST API. The `site_id` is the portion of the URL that follows the `/site/` in the URL. For example, "MarketingTeam" is the `site_id` in the following URL *MyServer*/#/site/**MarketingTeam**/projects. To specify the default site on Tableau Server, you can use an empty string **' '**.  For Tableau Online, you must provide a value for the `site_id`.
```
Comment (RussTheAerialist) [inline docs/docs/api-ref.md]: your empty string contains a space.

File: docs/docs/api-ref.md [hunk 6]
```diff
@@ -5,257 +5,3008 @@ layout: docs
...
-### Delete Project
+REST API: [Download Datasource](http://onlinehelp.tableau.com/current/api/rest_api/en-us/help.htm#REST/rest_api_ref.htm#Download_Datasource%3FTocPath%3DAPI%2520Reference%7C_____34){:target="_blank"}  
 
-Deletes a project by ID.
+**Parameters**
+
+Name | Description  
+:--- | :--- 
+`datasource_id` |  The identifier (`id`) for the the `DatasourceItem` that you want to download from the server. 
+`filepath` |  (Optional) Downloads the file to the location you specify. If no location is specified (the default is `Filepath=None`), the file is downloaded to the current working directory.
```
Comment (RussTheAerialist) [inline docs/docs/api-ref.md]: We should mention no_extract as a parameter

File: docs/docs/api-ref.md (hunk 1 above)
Comment (d45) [inline docs/docs/api-ref.md]: I'll take care of the quick fixes in this PR (against master).  I hear you about the ordering of the API references. I will look into that after I switch to development as base.

File: docs/docs/api-ref.md (hunk 3 above)
Comment (d45) [inline docs/docs/api-ref.md]: Removed the comments. Leftover from an experiment with kramdown and the toc_levels in the toc macro.

File: docs/docs/api-ref.md (hunk 4 above)
Comment (d45) [inline docs/docs/api-ref.md]: done

File: docs/docs/api-ref.md (hunk 5 above)
Comment (d45) [inline docs/docs/api-ref.md]: Good point.  I'll correct that and spell it out.

File: docs/docs/api-ref.md (hunk 6 above)
Comment (d45) [inline docs/docs/api-ref.md]: Added it. I'll add no_extract to workbooks.download too

File: docs/docs/api-ref.md [hunk 7]
```diff
@@ -310,7 +310,7 @@ Name | Description
 :--- | :--- 
 `datasource_id` |  The identifier (`id`) for the the `DatasourceItem` that you want to download from the server. 
 `filepath` |  (Optional) Downloads the file to the location you specify. If no location is specified (the default is `Filepath=None`), the file is downloaded to the current working directory. 
-
+`no_extract` | (Optional) Specifies whether to download the file without the extract. When the data source has an extract, if you set the parameter `no_extract=True`, the extract is not included. You can use this parameter to improve performance if you are downloading data sources that have large extracts. The default is to include the extract, if present (`no_extract=False`).
```
Comment (RussTheAerialist) [inline docs/docs/api-ref.md]: we should mention that this is REST v2.5+.

File: docs/docs/api-ref.md [hunk 8]
```diff
@@ -5,257 +5,3010 @@ layout: docs
...
+<br>  
 
-Signs you in to Tableau Server.
+### TableauAuth class
 
 ```py
-Auth.sign_in(authentication_object)
+TableauAuth(username, password, site_id='', user_id_to_impersonate=None)
 ```
+The `TableauAuth` class contains the attributes for the authentication resources. The `TableauAuth` class defines the information you can set in a  request or query from Tableau Server. The class members correspond to the attributes of a server request or response payload. To use this class, create a new instance, supplying user name, password, and site information if necessary, and pass the request object to the [Auth.sign_in](#auth.sign-in) method.
```
Comment (t8y8) [inline docs/docs/api-ref.md]: I'm a little skeptical calling it an authentication 'resource'. Not a big one, I'd actually just call it an object or credentials or something.

File: docs/docs/api-ref.md [hunk 9]
```diff
@@ -5,257 +5,3010 @@ layout: docs
...
+
+
+**Example**
 
 ```py
-new_site = TSC.SiteItem(name='Tableau', content_url='tableau', admin_mode=TSC.SiteItem.AdminMode.ContentAndUsers, user_quota=15, storage_quota=1000, disable_subscriptions=True)
-self.server.sites.create(new_site)
+import tableauserverclient as TSC
+
+# create a auth request
```
Comment (t8y8) [inline docs/docs/api-ref.md]: an auth

File: docs/docs/api-ref.md [hunk 10]
```diff
@@ -5,257 +5,3010 @@ layout: docs
...
+tableau_auth = TSC.TableauAuth('USERNAME', 'PASSWORD')
+
+# create an instance for your server
+server = TSC.Server('http://SERVER_URL')
+
+# call the sign-in method with the request
```
Comment (t8y8) [inline docs/docs/api-ref.md]: You call sign in with the object/resource/whatever-you-pick not the request, you're making a request

File: docs/docs/api-ref.md [hunk 11]
```diff
@@ -5,257 +5,3010 @@ layout: docs
...
+#### auth.sign out
 
 ```py
-Sites.get_by_id(id)
+auth.sign_out()
 ```
+Signs you out of the current session.
 
-### Get Sites
+The method takes care of invalidating the authentiction token. For more information, see [Sign in and Out](sign-in-out).
```
Comment (t8y8) [inline docs/docs/api-ref.md]: this method? or the method?

File: docs/docs/api-ref.md [hunk 12]
```diff
@@ -5,257 +5,3010 @@ layout: docs
...
+
+<br>
+<br>
+
+
+
+
+## Connections
+
+The connections for Tableau Server data sources and workbooks are represented by a `ConnectionItem` class.  You can call data source and view methods to query or update the connection information.  The `ConnectionCredentials` class represents the connection information you can update.
```
Comment (t8y8) [inline docs/docs/api-ref.md]: What does `and view methods` mean?

File: docs/docs/api-ref.md [hunk 13]
```diff
@@ -5,257 +5,3010 @@ layout: docs
...
+
+**Attributes**
+
+Attribute | Description
+:--- | :---
+`name`     | The username for the connection.
+`embed_password`  |  (Boolean) Determines whether to embed the passowrd (`True`) for the workbook or data source connection or not (`False`). 
+`password`  |  The password used for the connection.   
+`server_address`   |  The server address for the connection.   
+`server_port`   |  The port used by the server.
```
Comment (t8y8) [inline docs/docs/api-ref.md]: You say `for the connection` in the lines before to distinguish between the REST API server and the connection-server, I think that's a good idea and should happen on port too (or another way to clarify that)

File: docs/docs/api-ref.md [hunk 14]
```diff
@@ -5,257 +5,3010 @@ layout: docs
...
 
-### Create Project
+<br>
+<br>
 
-Creates a project for the given project item object.
+## Data sources
+
+Using the TSC library, you can get all the data sources on a site, or get the data sources for a specific project. 
+The data source resources for Tableau Server are defined in the `DatasourceItem` class. The class corresponds to the data source resources you can access using the Tableau Server REST API. For example, you can gather information about the name of the data source, its type, and connections, and the project it is associated with. The data source methods are based upon the endpoints for data sources in the REST API and operate on the `DatasourceItem` class.
```
Comment (t8y8) [inline docs/docs/api-ref.md]: three ands in a comma separated list

File: docs/docs/api-ref.md [hunk 15]
```diff
@@ -5,257 +5,3010 @@ layout: docs
...
 ```py
-Workbooks.delete(id)
+
+  import tableauserverclient as TSC
+  server = TSC.Server('http://SERVERURL')
+    
+  ...
+
+  project_id = '3a8b6148-493c-11e6-a621-6f3499394a39'
+  file_path = 'C:\\temp\\WorldIndicators.tde'
```
Comment (t8y8) [inline docs/docs/api-ref.md]: I prefer raw strings `r'C:\temp\WorldIndicators.tde'` -- style choice though

File: docs/docs/api-ref.md [hunk 16]
```diff
@@ -5,257 +5,3010 @@ layout: docs
...
+**Example**
 
 ```py
-Workbooks.populate_views(workbook_obj)
+# from server-client-python/test/test_datasource.py
+
+    ...
+
+    single_datasource = TSC.DatasourceItem('test', '1d0304cd-3796-429f-b815-7258370b9b74')
+    # need to specify the ID - this is a workaround.
```
Comment (t8y8) [inline docs/docs/api-ref.md]: I don't like workaround in the docs :( why is this necessary? /cc @RussTheAerialist 

That is, what's this sample showing, do we need to demonstrate it that way

File: docs/docs/api-ref.md [hunk 17]
```diff
@@ -5,257 +5,3010 @@ layout: docs
...
+
+### GroupItem class
 
 ```py
-Workbooks.populate_connections(workbook_obj)
+GroupItem(name)
 ```
 
-### Get Views for a Workbook
+The `GroupItem` class contains the members or attributes for the view resources on Tableau Server. The `GroupItem` class defines the information you can request or query from Tableau Server. The class members correspond to the attributes of a server request or response payload.
```
Comment (t8y8) [inline docs/docs/api-ref.md]: What are members?
I think it's just attributes :P

File: docs/docs/api-ref.md [hunk 18]
```diff
@@ -5,257 +5,3010 @@ layout: docs
...
+
+**Example**
+
+```py
+# Adding a user to a group
+# Using the second group on the site, aleady have all_groups
+# The id for Ian is '59a8a7b6-be3a-4d2d-1e9e-08a7b6b5b4ba'
+
+# add Ian to the second group
+  server.groups.add_user(all_groups[1], '59a8a7b6-be3a-4d2d-1e9e-08a7b6b5b4ba')
```
Comment (t8y8) [inline docs/docs/api-ref.md]: Whenever possible we should avoid the [subscript] notation in example because it's not obvious what it means. We can always pop it from the list and give it a name

File: docs/docs/api-ref.md [hunk 19]
```diff
@@ -5,257 +5,3010 @@ layout: docs
...
+#  server = TSC.Server('http://SERVERURL')
+
+
+# create a new instance with the group name
+  newgroup = TSC.GroupItem('My Group')
+
+# call the create method
+  newgroup = server.groups.create(newgroup)
+
+# print the names of the groups on the server
```
Comment (t8y8) [inline docs/docs/api-ref.md]: Without Pager this is only the first 100 groups

File: docs/docs/api-ref.md (hunk 16 above)
Comment (RussTheAerialist) [inline docs/docs/api-ref.md]: Yeah, I'm not sure what you are trying to show here?  In this case, that workaround was probably specifically to enable the test scenario and not a preferred way of doing something

Comment (RussTheAerialist) [inline docs/docs/api-ref.md]: Oh wait, if I look below.  In this case, we want to "get the datasource by id" from the server object before updating it.  This is definitely not what we want people to do.

File: docs/docs/api-ref.md [hunk 20]
```diff
@@ -5,257 +5,3008 @@ layout: docs
...
+<br>
+<br>
 
-* TOC
-{:toc}
 
 ## Authentication
 
-Source files: server/endpoint/auth_endpoint.py, models/tableau_auth.py
+You can use the TSC library to manage authentication, so you can sign in and sign out of Tableau Server and Tableau Online. The authentication resources for Tableau Server are defined in the `TableauAuth` class and they correspond to the authentication attributes you can access using the Tableau Server REST API.
```
Comment (d45) [inline docs/docs/api-ref.md]: No problem. It took me a while to get back on this.  I just made another commit and added the update_connections for workbooks. That is what caused the merge conflict.

File: docs/docs/api-ref.md (hunk 8 above)
Comment (d45) [inline docs/docs/api-ref.md]: Yes. I wasn't sure about that either. My wording here is pretty clumsy. I'll revise. Gotta run right now, but I will get back to it.

File: docs/docs/api-ref.md (hunk 20 above)
Comment (d45) [inline docs/docs/api-ref.md]: OK. Good point.  I managed to remove the managed and simplified it.

File: docs/docs/api-ref.md (hunk 8 above)
Comment (d45) [inline docs/docs/api-ref.md]: cut it down to this: 
The `TableauAuth` class defines the information you can set in a sign-in request. The class members correspond to the attributes of a server request or response payload. To use this class, create a new instance, supplying user name, password, and site information if necessary, and pass the request object to the [Auth.sign_in](#auth.sign-in) method.

File: docs/docs/api-ref.md (hunk 9 above)
Comment (d45) [inline docs/docs/api-ref.md]: Thanks!

File: docs/docs/api-ref.md (hunk 12 above)
Comment (d45) [inline docs/docs/api-ref.md]: That is an error.  Should be "call data source and workbook methods"

File: docs/docs/api-ref.md (hunk 13 above)
Comment (d45) [inline docs/docs/api-ref.md]: I updated "server_port" to say "the port used for the connection"

BTW I added some attributes to the ConnectionItem class (the username, password, etc.) that I missed earlier.  Also, I realized that the ConnectionCredentials is really just used when publishing workbooks and data sources. I didn't have that stated that way.

File: docs/docs/api-ref.md (hunk 14 above)
Comment (d45) [inline docs/docs/api-ref.md]: Um.. just two. But I changed it up. Just one now, but the sentence ended up with a bunch of "its"  :)

File: docs/docs/api-ref.md (hunk 7 above)
Comment (d45) [inline docs/docs/api-ref.md]: Good point. We should probably add the "available in REST version XX" for all of these. Or some.

File: docs/docs/api-ref.md (hunk 15 above)
Comment (d45) [inline docs/docs/api-ref.md]: I'll go with that. Thanks.

File: docs/docs/api-ref.md (hunk 16 above)
Comment (d45) [inline docs/docs/api-ref.md]: Yikes. Good catch. That is a placeholder example from the test code.

File: docs/docs/api-ref.md (hunk 11 above)
Comment (d45) [inline docs/docs/api-ref.md]: Changed to "The sigh_out() method ... "

File: docs/docs/api-ref.md (hunk 16 above)
Comment (d45) [inline docs/docs/api-ref.md]: replaced that example with this:

# get the data source item to update
  datasource = server.datasources.get_by_id('1a2a3b4b-5c6c-7d8d-9e0e-1f2f3a4a5b6b')
  
# do some updating 
  datasource.name = 'New Name'

# call the update method with the data source item
  updated_datasource = server.datasources.update(datasource)

File: docs/docs/api-ref.md (hunk 17 above)
Comment (d45) [inline docs/docs/api-ref.md]: yep. Fixed that. Also changed "view resources" to "group resources"  - copy and paste error.

File: docs/docs/api-ref.md (hunk 18 above)
Comment (d45) [inline docs/docs/api-ref.md]: Good point. I was just trying to get something to work and stopped at that. I will update these instances if I can.

File: docs/docs/api-ref.md (hunk 8 above)
Comment (t8y8) [inline docs/docs/api-ref.md]: beautiful

File: docs/docs/api-ref.md (hunk 18 above)
Comment (d45) [inline docs/docs/api-ref.md]: Rats! -- I goofed on some of these examples. We don't have a `groups.get_by_id()` method. 
Probably thinking of it as way to get around indexing [].  I guess I didn't test those.

File: docs/docs/api-ref.md (hunk 19 above)
Comment (d45) [inline docs/docs/api-ref.md]: added the 100 limit to the comment

**Comment by t8y8** [comment]
Text is looking great, only thing I have left (before maybe a final pass) is that I agree w/ @RussTheAerialist that the organization is a bit odd. Fillter/Sort inside the list alphabetically is very is jointed from their actual use.

**Comment by d45** [comment]
Thanks for digging into this. Your review has been super helpful and has
uncovered a lot of issues. I finally ran spell check and took care of some
ridiculous 'errrors".

About the organization. I agree that it doesn't make sense if you are
looking at it from a "how do I do xx" perspective. Ideally, the modules
would be separate files, so it would be easy to sort them around in a more
logical manner (models, endpoints,etc).  We don't have that at the moment.
Because it is a long, long file (like 70 pages in Word) I think the
alphabetical approach is only way of making it almost useful for finding
stuff and scrolling around. But just barely.  Breaking the file up into
sections makes sense.

Anyway... what we really need is to add more to the "Getting Started"
pages. That way we could described how the library is organized and what
you need to do to begin using it (as Russell said, starting with the server
and working out to the resources. If we do that, how the reference pages
are organized won't matter as much.

Oh, it looks like the merge conflict is just with the workbooks update
method. I added that, so I think that we should be good.
